File: cell/src/api/firestore.ts
import { initializeApp } from 'firebase/app';
import {
  addDoc,
  collection,
  deleteDoc,
  doc,
  getDoc,
  getDocs,
  getFirestore,
  onSnapshot,
  query,
  serverTimestamp,
  updateDoc,
  where
} from 'firebase/firestore';
import { firebaseConfig } from '../config/firebase';
import { Message, Personality, Thread } from '../types';

// Initialize Firebase
const app = initializeApp(firebaseConfig);
const db = getFirestore(app);

// Collections
const MESSAGES_COLLECTION = 'messages';
const THREADS_COLLECTION = 'threads';

/**
 * Save a message to Firestore
 */
export async function saveMessage(message: Omit<Message, 'id'>): Promise<string> {
  try {
    const docRef = await addDoc(collection(db, MESSAGES_COLLECTION), {
      ...message,
      createdAt: serverTimestamp(),
    });
    return docRef.id;
  } catch (error) {
    console.error('Error saving message:', error);
    throw new Error('Failed to save message');
  }
}

/**
 * Get messages for a thread
 */
export async function getMessages(threadId: string, ownerId: string): Promise<Message[]> {
  try {
    const messagesRef = collection(db, MESSAGES_COLLECTION);
    // Simplified query without orderBy to avoid index requirement
    const q = query(
      messagesRef,
      where('threadId', '==', threadId),
      where('ownerId', '==', ownerId)
    );
    
    const querySnapshot = await getDocs(q);
    const messages: Message[] = [];
    
    querySnapshot.forEach((doc) => {
      const data = doc.data();
      messages.push({
        id: doc.id,
        ...data,
        createdAt: data.createdAt?.toMillis() || Date.now(),
      } as Message);
    });
    
    // Sort messages by createdAt on the client side
    messages.sort((a, b) => a.createdAt - b.createdAt);
    
    return messages;
  } catch (error) {
    console.error('Error getting messages:', error);
    throw new Error('Failed to get messages');
  }
}

/**
 * Update message status
 */
export async function updateMessageStatus(
  messageId: string, 
  status: Message['status'], 
  error?: string | null
): Promise<void> {
  try {
    const messageRef = doc(db, MESSAGES_COLLECTION, messageId);
    const updateData: any = {
      status,
      updatedAt: serverTimestamp(),
    };
    
    // Only include error field if it's not undefined
    if (error !== undefined) {
      updateData.error = error;
    }
    
    await updateDoc(messageRef, updateData);
  } catch (error) {
    console.error('Error updating message status:', error);
    throw new Error('Failed to update message status');
  }
}

/**
 * Update message payload (for transcript, improvement, etc.)
 */
export async function updateMessagePayload(
  messageId: string, 
  payloadUpdates: Partial<Message['payload']>
): Promise<void> {
  try {
    const messageRef = doc(db, MESSAGES_COLLECTION, messageId);
    
    // Get current message to merge with existing payload
    const currentDoc = await getDoc(messageRef);
    let currentPayload = {};
    
    if (currentDoc.exists()) {
      const currentData = currentDoc.data();
      currentPayload = currentData.payload || {};
    }
    
    // Merge existing payload with updates
    const mergedPayload = {
      ...currentPayload,
      ...payloadUpdates
    };
    
    await updateDoc(messageRef, {
      payload: mergedPayload,
      updatedAt: serverTimestamp(),
    });
  } catch (error) {
    console.error('Error updating message payload:', error);
    throw new Error('Failed to update message payload');
  }
}

/**
 * Delete a message
 */
export async function deleteMessage(messageId: string): Promise<void> {
  try {
    const messageRef = doc(db, MESSAGES_COLLECTION, messageId);
    await deleteDoc(messageRef);
  } catch (error) {
    console.error('Error deleting message:', error);
    throw new Error('Failed to delete message');
  }
}

/**
 * Save a thread to Firestore
 */
export async function saveThread(thread: Omit<Thread, 'id'>): Promise<string> {
  try {
    const docRef = await addDoc(collection(db, THREADS_COLLECTION), {
      ...thread,
      createdAt: serverTimestamp(),
      updatedAt: serverTimestamp(),
    });
    return docRef.id;
  } catch (error) {
    console.error('Error saving thread:', error);
    throw new Error('Failed to save thread');
  }
}

/**
 * Get a single thread by ID
 */
export async function getThread(threadId: string): Promise<Thread | null> {
  try {
    const threadRef = doc(db, THREADS_COLLECTION, threadId);
    const threadSnap = await getDoc(threadRef);
    
    if (threadSnap.exists()) {
      const data = threadSnap.data();
      return {
        id: threadSnap.id,
        ...data,
        createdAt: data.createdAt?.toMillis() || Date.now(),
      } as Thread;
    }
    
    return null;
  } catch (error) {
    console.error('Error getting thread:', error);
    throw error;
  }
}

/**
 * Get threads for a user
 */
export async function getThreads(ownerId: string): Promise<Thread[]> {
  try {
    const threadsRef = collection(db, THREADS_COLLECTION);
    // Sem filtro: retorna todas as threads (atenção à segurança nas regras do Firestore)
    const q = query(threadsRef);
    const querySnapshot = await getDocs(q);
    const threads: Thread[] = [];
    
    querySnapshot.forEach((doc) => {
      const data = doc.data();
      threads.push({
        id: doc.id,
        ...data,
        createdAt: data.createdAt?.toMillis() || Date.now(),
        updatedAt: data.updatedAt?.toMillis() || Date.now(),
      } as Thread);
    });
    
    // Sort threads by updatedAt on the client side
    threads.sort((a, b) => b.updatedAt - a.updatedAt);
    
    return threads;
  } catch (error) {
    console.error('Error getting threads:', error);
    throw new Error('Failed to get threads');
  }
}

/**
 * Update thread
 */
export async function updateThread(threadId: string, updates: Partial<Thread>): Promise<void> {
  try {
    const threadRef = doc(db, THREADS_COLLECTION, threadId);
    await updateDoc(threadRef, {
      ...updates,
      updatedAt: serverTimestamp(),
    });
  } catch (error) {
    console.error('Error updating thread:', error);
    throw new Error('Failed to update thread');
  }
}

/**
 * Delete a thread
 */
export async function deleteThread(threadId: string): Promise<void> {
  try {
    const threadRef = doc(db, THREADS_COLLECTION, threadId);
    await deleteDoc(threadRef);
  } catch (error) {
    console.error('Error deleting thread:', error);
    throw new Error('Failed to delete thread');
  }
}

/**
 * Subscribe to real-time updates for messages in a thread
 */
export function subscribeToMessages(
  threadId: string, 
  deviceId: string, 
  onUpdate: (messages: Message[]) => void
): () => void {
  const messagesRef = collection(db, MESSAGES_COLLECTION);
  const q = query(
    messagesRef,
    where('threadId', '==', threadId),
    where('deviceId', '==', deviceId)
  );
  
  const unsubscribe = onSnapshot(q, (querySnapshot) => {
    const messages: Message[] = [];
    
    querySnapshot.forEach((doc) => {
      const data = doc.data();
      messages.push({
        id: doc.id,
        ...data,
        createdAt: data.createdAt?.toMillis() || Date.now(),
      } as Message);
    });
    
    // Sort messages by creation time
    messages.sort((a, b) => a.createdAt - b.createdAt);
    
    onUpdate(messages);
  }, (error) => {
    console.error('Error in realtime subscription:', error);
  });
  
  return unsubscribe;
}

// Personality functions
export const savePersonality = async (personality: Omit<Personality, 'id' | 'createdAt' | 'updatedAt'>): Promise<string> => {
  try {
    const personalityData = {
      ...personality,
      createdAt: Date.now(),
      updatedAt: Date.now()
    };
    
    const docRef = await addDoc(collection(db, 'personalities'), personalityData);
    console.log('Personality saved with ID:', docRef.id);
    return docRef.id;
  } catch (error) {
    console.error('Error saving personality:', error);
    throw error;
  }
};

export const updatePersonality = async (id: string, updates: Partial<Personality>): Promise<void> => {
  try {
    const personalityRef = doc(db, 'personalities', id);
    const updateData = {
      ...updates,
      updatedAt: Date.now()
    };
    
    await updateDoc(personalityRef, updateData);
    console.log('Personality updated:', id);
  } catch (error) {
    console.error('Error updating personality:', error);
    throw error;
  }
};

export const deletePersonality = async (id: string): Promise<void> => {
  try {
    const personalityRef = doc(db, 'personalities', id);
    await deleteDoc(personalityRef);
    console.log('Personality deleted:', id);
  } catch (error) {
    console.error('Error deleting personality:', error);
    throw error;
  }
};

export const getPersonalities = async (): Promise<Personality[]> => {
  try {
    const querySnapshot = await getDocs(collection(db, 'personalities'));
    const personalities = querySnapshot.docs.map(doc => ({
      id: doc.id,
      ...doc.data()
    })) as Personality[];
    
    console.log('Fetched personalities:', personalities.length);
    return personalities;
  } catch (error) {
    console.error('Error fetching personalities:', error);
    throw error;
  }
};

File: desktop/src/firestore.py
"""
Módulo para integração com Firestore
Implementação idêntica ao mobile
"""

import logging
from typing import List, Dict, Any, Optional, Callable
from datetime import datetime
from google.cloud.firestore_v1.base_query import FieldFilter
from google.cloud import firestore

from .types import Message, Thread, MessageKind, MessageStatus, MessageSource, message_to_dict, message_from_dict, thread_to_dict, thread_from_dict
from .device_id import get_or_create_device_id

logger = logging.getLogger(__name__)

class FirestoreManager:
    """Gerenciador de integração com Firestore - igual ao mobile"""
    
    def __init__(self, auth_manager=None):
        # Inicializar Firebase diretamente
        from google.cloud import firestore
        self.db = firestore.Client()
        
    def save_message(self, message: Message) -> str:
        """
        Salvar mensagem no Firestore - igual ao mobile
        """
        if not self.db:
            raise Exception("Firebase não inicializado")
            
        try:
            message_dict = message_to_dict(message)
            # Remover id do dict para criar novo documento
            message_id = message_dict.pop('id')
            
            doc_ref = self.db.collection('messages').add(message_dict)
            return doc_ref[1].id  # ID do documento criado
            
        except Exception as e:
            logger.error(f"Erro ao salvar mensagem: {e}")
            raise Exception("Falha ao salvar mensagem")
            
    def get_messages(self, thread_id: str, owner_id: str = None) -> List[Message]:
        """
        Obter mensagens de uma thread (conversas globais)
        """
        if not self.db:
            return []
            
        try:
            messages_ref = self.db.collection('messages')
            # Buscar mensagens apenas por threadId (sem filtro de ownerId)
            query = messages_ref.where('threadId', '==', thread_id)
            docs = query.stream()
            
            messages = []
            for doc in docs:
                data = doc.to_dict()
                data['id'] = doc.id
                # Converter timestamp do Firestore para timestamp Unix
                if 'createdAt' in data and hasattr(data['createdAt'], 'timestamp'):
                    data['createdAt'] = int(data['createdAt'].timestamp() * 1000)
                messages.append(message_from_dict(data))
                
            # Ordenar por data de criação
            messages.sort(key=lambda x: x.createdAt)
            
            logger.info(f"Encontradas {len(messages)} mensagens na thread {thread_id}")
            return messages
            
        except Exception as e:
            logger.error(f"Erro ao obter mensagens da thread {thread_id}: {e}")
            return []
            
    def update_message_status(self, message_id: str, status: MessageStatus, error: Optional[str] = None) -> None:
        """
        Atualizar status da mensagem - igual ao mobile
        """
        if not self.db:
            raise Exception("Firebase não inicializado")
            
        try:
            message_ref = self.db.collection('messages').document(message_id)
            update_data = {
                'status': status.value,
                'updatedAt': firestore.SERVER_TIMESTAMP
            }
            
            if error is not None:
                update_data['error'] = error
                
            message_ref.update(update_data)
            
        except Exception as e:
            logger.error(f"Erro ao atualizar status da mensagem {message_id}: {e}")
            raise Exception("Falha ao atualizar status da mensagem")
            
    def update_message_payload(self, message_id: str, payload_updates: Dict[str, Any]) -> None:
        """
        Atualizar payload da mensagem - igual ao mobile
        """
        if not self.db:
            raise Exception("Firebase não inicializado")
            
        try:
            message_ref = self.db.collection('messages').document(message_id)
            
            # Obter mensagem atual para merge com payload existente
            current_doc = message_ref.get()
            current_payload = {}
            
            if current_doc.exists:
                current_data = current_doc.to_dict()
                current_payload = current_data.get('payload', {})
            
            # Merge com atualizações
            merged_payload = {**current_payload, **payload_updates}
            
            message_ref.update({
                'payload': merged_payload,
                'updatedAt': firestore.SERVER_TIMESTAMP
            })
            
        except Exception as e:
            logger.error(f"Erro ao atualizar payload da mensagem {message_id}: {e}")
            raise Exception("Falha ao atualizar payload da mensagem")
            
    def delete_message(self, message_id: str) -> None:
        """
        Deletar mensagem - igual ao mobile
        """
        if not self.db:
            raise Exception("Firebase não inicializado")
            
        try:
            message_ref = self.db.collection('messages').document(message_id)
            message_ref.delete()
            
        except Exception as e:
            logger.error(f"Erro ao deletar mensagem {message_id}: {e}")
            raise Exception("Falha ao deletar mensagem")
            
    def save_thread(self, thread: Thread) -> str:
        """
        Salvar thread - igual ao mobile
        """
        if not self.db:
            raise Exception("Firebase não inicializado")
            
        try:
            thread_dict = thread_to_dict(thread)
            # Remover id do dict para criar novo documento
            thread_id = thread_dict.pop('id')
            
            doc_ref = self.db.collection('threads').add(thread_dict)
            return doc_ref[1].id  # ID do documento criado
            
        except Exception as e:
            logger.error(f"Erro ao salvar thread: {e}")
            raise Exception("Falha ao salvar thread")
            
    def get_thread(self, thread_id: str) -> Optional[Thread]:
        """
        Obter thread por ID - igual ao mobile
        """
        if not self.db:
            return None
            
        try:
            thread_ref = self.db.collection('threads').document(thread_id)
            thread_doc = thread_ref.get()
            
            if thread_doc.exists:
                data = thread_doc.to_dict()
                data['id'] = thread_doc.id
                # Converter timestamp do Firestore para timestamp Unix
                if 'createdAt' in data and hasattr(data['createdAt'], 'timestamp'):
                    data['createdAt'] = int(data['createdAt'].timestamp() * 1000)
                if 'updatedAt' in data and hasattr(data['updatedAt'], 'timestamp'):
                    data['updatedAt'] = int(data['updatedAt'].timestamp() * 1000)
                return thread_from_dict(data)
            
            return None
            
        except Exception as e:
            logger.error(f"Erro ao obter thread {thread_id}: {e}")
            return None
            
    def get_threads(self, owner_id: str = None) -> List[Thread]:
        """
        Obter todas as threads (conversas globais)
        """
        if not self.db:
            return []
            
        try:
            threads_ref = self.db.collection('threads')
            # Buscar todas as threads sem filtro de ownerId
            docs = threads_ref.stream()
            threads = []
            for doc in docs:
                data = doc.to_dict()
                data['id'] = doc.id
                # Converter timestamp do Firestore para timestamp Unix
                if 'createdAt' in data and hasattr(data['createdAt'], 'timestamp'):
                    data['createdAt'] = int(data['createdAt'].timestamp() * 1000)
                if 'updatedAt' in data and hasattr(data['updatedAt'], 'timestamp'):
                    data['updatedAt'] = int(data['updatedAt'].timestamp() * 1000)
                threads.append(thread_from_dict(data))
                
            # Ordenar por data de atualização (mais recente primeiro)
            threads.sort(key=lambda x: x.updatedAt, reverse=True)
            
            logger.info(f"Encontradas {len(threads)} threads globais")
            return threads
            
        except Exception as e:
            logger.error(f"Erro ao obter threads: {e}")
            return []
            
    def update_thread(self, thread_id: str, updates: Dict[str, Any]) -> None:
        """
        Atualizar thread - igual ao mobile
        """
        if not self.db:
            raise Exception("Firebase não inicializado")
            
        try:
            thread_ref = self.db.collection('threads').document(thread_id)
            update_data = {
                **updates,
                'updatedAt': firestore.SERVER_TIMESTAMP
            }
            thread_ref.update(update_data)
            
        except Exception as e:
            logger.error(f"Erro ao atualizar thread {thread_id}: {e}")
            raise Exception("Falha ao atualizar thread")
            
    def delete_thread(self, thread_id: str) -> None:
        """
        Deletar thread - igual ao mobile
        """
        if not self.db:
            raise Exception("Firebase não inicializado")
            
        try:
            thread_ref = self.db.collection('threads').document(thread_id)
            thread_ref.delete()
            
        except Exception as e:
            logger.error(f"Erro ao deletar thread {thread_id}: {e}")
            raise Exception("Falha ao deletar thread")
            
    def subscribe_to_messages(self, thread_id: str, device_id: str, on_update: Callable[[List[Message]], None]) -> Callable[[], None]:
        """
        Inscrever-se para atualizações em tempo real - igual ao mobile
        """
        if not self.db:
            return lambda: None
            
        try:
            messages_ref = self.db.collection('messages')
            query = messages_ref.where('threadId', '==', thread_id).where('deviceId', '==', device_id)
            
            def on_snapshot(query_snapshot, changes, read_time):
                messages = []
                for doc in query_snapshot:
                    data = doc.to_dict()
                    data['id'] = doc.id
                    # Converter timestamp do Firestore para timestamp Unix
                    if 'createdAt' in data and hasattr(data['createdAt'], 'timestamp'):
                        data['createdAt'] = int(data['createdAt'].timestamp() * 1000)
                    messages.append(message_from_dict(data))
                
                # Ordenar por data de criação
                messages.sort(key=lambda x: x.createdAt)
                on_update(messages)
            
            unsubscribe = query.on_snapshot(on_snapshot)
            return unsubscribe
            
        except Exception as e:
            logger.error(f"Erro ao se inscrever em mensagens da thread {thread_id}: {e}")
            return lambda: None